File: AI_Agents/api/tools/jira_tool.js
const axios = require('axios');

class JiraTool {
    constructor(url, email, token) {
        this.url = (url || '').replace(/\/$/, '');
        this.email = email;
        this.token = token;

        const authString = Buffer.from(`${this.email}:${this.token}`).toString('base64');
        this.headers = {
            'Authorization': `Basic ${authString}`,
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        };
    }

    async testConnection() {
        try {
            const response = await axios.get(`${this.url}/rest/api/3/myself`, {
                headers: this.headers,
                timeout: 10000
            });
            return response.data;
        } catch (error) {
            const errStr = error.response ? `${error.response.status} - ${JSON.stringify(error.response.data)}` : error.message;
            throw new Error(`Jira Connection Test failed: ${errStr}`);
        }
    }

    async fetchIssue(issueId) {
        try {
            const response = await axios.get(`${this.url}/rest/api/3/issue/${issueId}`, {
                headers: this.headers,
                timeout: 15000
            });
            const data = response.data;
            const fields = data.fields || {};
            
            return {
                id: data.key,
                summary: fields.summary || 'No Summary',
                description: this.extractTextFromAdf(fields.description) || 'No Description provided.',
                status: fields.status?.name || 'Unknown',
                project: fields.project?.name || 'Unknown Project'
            };
        } catch (error) {
            const errStr = error.response ? `${error.response.status} - ${JSON.stringify(error.response.data)}` : error.message;
            throw new Error(`Jira Fetch error: ${errStr}`);
        }
    }

    async searchIssues(jql, maxResults = 5) {
        try {
            const response = await axios.get(`${this.url}/rest/api/3/search/jql`, {
                headers: this.headers,
                params: {
                    jql: jql,
                    maxResults: maxResults,
                    fields: 'summary,description,status,project'
                },
                timeout: 15000
            });
            
            const issues = response.data.issues || [];
            return issues.map(data => ({
                id: data.key,
                summary: data.fields?.summary || 'No Summary',
                description: this.extractTextFromAdf(data.fields?.description) || 'No Description provided.',
                status: data.fields?.status?.name || 'Unknown',
                project: data.fields?.project?.name || 'Unknown Project'
            }));
        } catch (error) {
            const errStr = error.response ? `${error.response.status} - ${JSON.stringify(error.response.data)}` : error.message;
            throw new Error(`Jira Search error: ${errStr}`);
        }
    }

    // Helper to extract plain text from Jira's Atlassian Document Format (ADF)
    extractTextFromAdf(description) {
        if (!description) return "";
        if (typeof description === 'string') return description;
        
        try {
            if (description.content && Array.isArray(description.content)) {
                let text = "";
                for (const block of description.content) {
                    if (block.content && Array.isArray(block.content)) {
                        for (const inline of block.content) {
                            if (inline.text) text += inline.text;
                        }
                    }
                    text += "\n";
                }
                return text.trim();
            }
        } catch (e) {
            console.error("ADF Extraction Error:", e);
        }
        return JSON.stringify(description);
    }
}

module.exports = JiraTool;

File: AI_Agents/tools/jira_tool.py
import requests
from typing import Dict, Any

class JiraTool:
    def __init__(self, url: str, email: str, token: str):
        self.url = url.rstrip("/")
        self.auth = (email, token)
        self.headers = {
            "Accept": "application/json",
            "Content-Type": "application/json"
        }

    def test_connection(self) -> Dict[str, Any]:
        """Verify Jira connection and return user data."""
        api_url = f"{self.url}/rest/api/3/myself"
        response = requests.get(api_url, auth=self.auth, headers=self.headers, timeout=10)
        if response.status_code != 200:
            raise Exception(f"Jira Auth failed: {response.status_code}")
        return response.json()

    def search_issues(self, jql: str, max_results: int = 5) -> list:
        """Search for issues using JQL."""
        # Migrate to the latest /search/jql API as per Atlassian CHANGE-2046
        api_url = f"{self.url}/rest/api/3/search/jql"
        params = {
            "jql": jql,
            "maxResults": max_results,
            "fields": "summary,description,status,project"
        }
        response = requests.get(api_url, auth=self.auth, headers=self.headers, params=params, timeout=15)
        
        if response.status_code != 200:
            raise Exception(f"Jira Search error: {response.status_code} - {response.text}")
            
        issues = response.json().get("issues", [])
        results = []
        for issue in issues:
            fields = issue.get("fields", {})
            results.append({
                "id": issue.get("key"),
                "summary": fields.get("summary", ""),
                "description": self._parse_adf(fields.get("description", {})),
                "status": fields.get("status", {}).get("name", ""),
                "project": fields.get("project", {}).get("name", "")
            })
        return results

    def fetch_issue(self, issue_id: str) -> Dict[str, Any]:
        """Fetch issue details from Jira."""
        api_url = f"{self.url}/rest/api/3/issue/{issue_id}"
        response = requests.get(api_url, auth=self.auth, headers=self.headers, timeout=15)
        
        if response.status_code != 200:
            raise Exception(f"Jira API error: {response.status_code} - {response.text}")
            
        data = response.json()
        fields = data.get("fields", {})
        
        return {
            "id": issue_id,
            "summary": fields.get("summary", ""),
            "description": self._parse_adf(fields.get("description", {})),
            "status": fields.get("status", {}).get("name", ""),
            "project": fields.get("project", {}).get("name", ""),
            "raw": data # Keep raw for advanced context
        }

    def _parse_adf(self, adf_content: Dict[str, Any]) -> str:
        """Simple Atlassian Document Format parser to extract text."""
        if not adf_content or "content" not in adf_content:
            return ""
        
        text_parts = []
        for block in adf_content["content"]:
            if block.get("type") == "paragraph":
                for item in block.get("content", []):
                    if item.get("type") == "text":
                        text_parts.append(item.get("text", ""))
        
        return "\n".join(text_parts)
